homeassistant/components/mqtt/device_automation.py

```diff
@@ -0,0 +1,227 @@
+"""Provides device automations for mqtt."""
+import logging
+
+import attr
+import voluptuous as vol
+
+from homeassistant.components import mqtt
+import homeassistant.components.automation.mqtt as automation_mqtt
+from homeassistant.const import CONF_DEVICE_ID, CONF_DOMAIN, CONF_NAME, CONF_PLATFORM
+import homeassistant.helpers.config_validation as cv
+from homeassistant.core import callback
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from . import (
+    ATTR_DISCOVERY_HASH,
+    CONF_CONNECTIONS,
+    CONF_DEVICE,
+    CONF_IDENTIFIERS,
+    CONF_PAYLOAD,
+    CONF_QOS,
+    DOMAIN,
+)
+from .discovery import MQTT_DISCOVERY_NEW, MQTT_DISCOVERY_UPDATED, clear_discovery_hash
+
+_LOGGER = logging.getLogger(__name__)
+
+# mypy: allow-untyped-defs, no-check-untyped-defs
+
+CONF_ENCODING = "encoding"
+DEFAULT_ENCODING = "utf-8"
+CONF_TOPIC = "topic"
+
+CONF_AUTOMATION_TYPE = "automation_type"
+AUTOMATION_TYPE_TRIGGER = "trigger"
+AUTOMATION_TYPES = [AUTOMATION_TYPE_TRIGGER]
+AUTOMATION_TYPES_SCHEMA = vol.In(AUTOMATION_TYPES)
+
+MQTT_TRIGGER = {
+    # Trigger when MQTT message is received
+    CONF_PLATFORM: "device",
+    CONF_DOMAIN: DOMAIN,
+}
+
+TRIGGER_SCHEMA = vol.All(
+    vol.Schema(
+        {
+            vol.Required(CONF_PLATFORM): "device",
+            vol.Required(CONF_DOMAIN): DOMAIN,
+            vol.Optional(CONF_DEVICE_ID): str,
+            vol.Required(CONF_TOPIC): mqtt.valid_subscribe_topic,
+            vol.Optional(CONF_PAYLOAD): cv.string,
+            vol.Optional(CONF_ENCODING, default=DEFAULT_ENCODING): cv.string,
+            vol.Optional(CONF_NAME): cv.string,
+        }
+    )
+)
+
+PLATFORM_SCHEMA = mqtt.MQTT_BASE_PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_AUTOMATION_TYPE): AUTOMATION_TYPES_SCHEMA,
```

Comment:
The idea is to allow other MQTT device automations (action, condition) without entities, but maybe that does not make sense and only trigger is needed?